extensions/guacamole-auth-sso/modules/guacamole-auth-sso-base/src/main/java/org/apache/guacamole/auth/sso/AuthenticationSessionManager.java

```diff
@@ -97,9 +149,21 @@ public String generateInvalid() {
      *     value was returned by defer().
      */
     public T resume(String identifier) {
-
         if (identifier != null) {
-            T session = sessions.remove(identifier);
+             T session = sessions.get(identifier);
+
+            // If this session is already pending, invalidate and return null
+            if (pendingSessions.contains(identifier)) {
+                invalidateSession(identifier);
+                return null;
+            }
+
+            // Otherwise, mark the session as pending. NOTE: Unless explicitly
+            // removed from pending status via a call to reactivateSession(),
+            // the next attempt to resume this session will fail
+            else
+                pendingSessions.add(identifier);
```

Comment:
Though the underlying collections are threadsafe, this access pattern is not. The intent appears to be that, given two separate calls to `resume("someValue")`, the first call succeeds and returns a non-null session while the second call returns `null`, but this may not happen. Consider:

| Thread 1                                              | Thread 2                      |
| ----------------------------------------------------- | ------------------------------------------------------ |
|`resume("someValue")`||
|                                                       | `resume("someValue")`                                  |
| `if (identifier != null) {` (true)                    |                              |
|                                                       | `if (identifier != null) {` (true)  |
| `T session = sessions.get(identifier);`               |                                                        |
|                                                       | `T session = sessions.get(identifier);`                |
| `if (pendingSessions.contains(identifier)) {` (false) |                                                        |
|                                                       |  `if (pendingSessions.contains(identifier)) {` (false) |
| `pendingSessions.add(identifier);`                    |  |
|                                                       | `pendingSessions.add(identifier);`                   |
| `if (session != null && session.isValid())` (true)    |  |
|                                                       | `if (session != null && session.isValid())` (true)  |
| `return session;`                                     |  |
|                                                       | `return session;`   |

Reply:
Oops, my bad. Yeah that's not going to work. I've updated it and this issue should be fixed now.